axiom/optimizer/ToGraph.h

```diff
@@ -61,17 +61,20 @@ struct ExprDedupHasher {
 using FunctionDedupMap =
     folly::F14FastMap<ExprDedupKey, ExprCP, ExprDedupHasher>;
 
-struct VariantPtrHasher {
-  size_t operator()(const std::shared_ptr<const velox::Variant>& value) const {
-    return value->hash();
+struct TypedVariant {
+  const velox::Type* type;
+  std::shared_ptr<const velox::Variant> value;
+};
+
+struct TypedVariantHasher {
+  size_t operator()(const TypedVariant& value) const {
+    return velox::bits::hashMix(value.type->hashKind(), value.value->hash());
   }
 };
 
-struct VariantPtrComparer {
-  bool operator()(
-      const std::shared_ptr<const velox::Variant>& left,
-      const std::shared_ptr<const velox::Variant>& right) const {
-    return *left == *right;
+struct TypedVariantComparer {
+  bool operator()(const TypedVariant& left, const TypedVariant& right) const {
+    return *left.type == *right.type && *left.value == *right.value;
```

Comment:
```suggestion
    return left.type == right.type && *left.value == *right.value;
```
They're deduplicated

Reply:
Good points. Updated. Thanks.